packages/extension-ui/src/hooks/useLedger.ts

```diff
@@ -0,0 +1,133 @@
+  ledger: Ledger | null;
+  refresh: () => void;
+  warning: string | null;
+}
+
+function getNetwork (genesis: string): Network | undefined {
+  return ledgerChains.find(({ genesisHash }) => genesisHash[0] === genesis);
+}
+
+function retrieveLedger (genesis: string): Ledger {
+  let ledger: Ledger | null = null;
+
+  const { isLedgerCapable } = getState();
+
+  assert(isLedgerCapable, 'Incompatible browser, only Chrome is supporter');
+
+  const def = getNetwork(genesis);
+
+  assert(def, `Unable to find supported chain for ${genesis}`);
+
+  ledger = new Ledger('webusb', def.network);
+
+  return ledger;
+}
+
+function getState (): StateBase {
+  const isLedgerCapable = !!(window as unknown as { USB?: unknown }).USB;
+
+  return {
+    isLedgerCapable,
+    isLedgerEnabled: isLedgerCapable && uiSettings.ledgerConn !== 'none'
+  };
+}
+
+export function useLedger (genesis?: string | null, accountIndex = 0, addressOffset = 0): State {
+  const [isLoading, setIsLoading] = useState(false);
+  const [isLocked, setIsLocked] = useState(false);
+  const [refreshLock, setRefreshLock] = useState(false);
+  const [warning, setWarning] = useState<string | null>(null);
+  const [error, setError] = useState<string | null>(null);
+  const [address, setAddress] = useState<string | null>(null);
+  const { t } = useTranslation();
+  const ledger = useMemo(() => {
+    setIsLocked(false);
+    setRefreshLock(false);
+
+    // this trick allows to refresh the ledger on demand
+    // when it is shown as locked and the user has actually
+    // unlocked it, which we can't know.
+    if (refreshLock || genesis) {
+      if (!genesis) {
+        return null;
+      }
+
+      return retrieveLedger(genesis);
+    }
+
+    return null;
+  }, [genesis, refreshLock]);
+
+  useEffect(() => {
+    if (!ledger || !genesis) {
+      setAddress(null);
+
+      return;
+    }
+
+    setIsLoading(true);
+    setError(null);
+    setWarning(null);
+
+    // sometimes, somehow addIndex is perceived as a string
+    ledger.getAddress(false, Number(accountIndex), Number(addressOffset))
+      .then((res) => {
+        setIsLoading(false);
+        setAddress(res.address);
+      }).catch((e: Error) => {
+        setIsLoading(false);
+        const { network } = getNetwork(genesis) || { network: 'unknown network' };
+
+        const warningMessage = e.message.includes('Code: 26628')
+          ? t<string>('Is your ledger locked?')
+          : null;
+
+        const errorMessage = e.message.includes('App does not seem to be open')
+          ? t<string>('App "{{network}}" does not seem to be open', { replace: { network } })
+          : e.message;
+
+        setIsLocked(true);
+        setWarning(warningMessage);
+        setError(t<string>(
+          'Ledger error: {{errorMessage}}',
+          { replace: { errorMessage } }
+        ));
+        console.error(e);
+        setAddress(null);
+      });
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
This dep array should contain `t`, the translation function. But if it has it it, it triggers a useless re-rendering of this component, it's visible at ~0:28 in the demo [here](https://github.com/polkadot-js/extension/pull/624#issuecomment-763747721) when taking a look at the button loading status. I can't pin point why. Ever had such a problem?

Reply:
Weird. t should be relatively "stable".

Reply:
Super weird, I've tried many things like removing the `{ replace : {}}`, take the `t` to another function with `useCallback`, it keeps re-rendering whatever I put it in. And because `useTranslation` is in a hook, I can't memoize it. Maybe it's a side effect of `web2usb`? it refreshes when the `ledger` gets available.

Even weirder, it didn't happen when it wasn't in a custom hook, with the same code [here](https://github.com/polkadot-js/extension/blob/fb0b7671642b44c52f302496090f8372dba5822a/packages/extension-ui/src/Popup/ImportLedger.tsx#L96).

Reply:
Just add a note before that comment as to "why". (We will wonder a month from now)

However, leaving out t should be very safe I believe. So should not have any side-effects.

Reply:
```suggestion
  // If the dependency array is exhaustive, with t, the translation function, it
  // triggers a useless re-render when ledger device is connected.
  // eslint-disable-next-line react-hooks/exhaustive-deps
```